#**Question 1**

Design a convolutional neural network in Keras of at least 10 convolutional layers. Use the MNIST dataset  for  evaluation. 

In [16]:
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from tensorflow.keras.optimizers import SGD, Adam, RMSprop
from keras.utils import np_utils

num_classes = 10 
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_test /= 255
x_train = x_train[:, np.newaxis, :, :]
x_test = x_test[:, np.newaxis, :, :]
print(x_train.shape, 'train samples')
print(x_test.shape, 'test samples')
Y_train = np_utils.to_categorical(y_train, num_classes)
Y_test = np_utils.to_categorical(y_test, num_classes)

(60000, 1, 28, 28) train samples
(10000, 1, 28, 28) test samples


In [17]:
import keras
from keras import layers
from keras.layers.convolutional import Conv2D
from keras.layers import Flatten
from keras.layers.convolutional import MaxPooling2D
def get_model1(input_shape, num_classes):
    model = keras.Sequential(
        [
            layers.Conv2D(3, kernel_size=3, padding="same", input_shape=input_shape),
            layers.Conv2D(4, kernel_size=3, padding="same"),
            layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
            layers.Conv2D(5, kernel_size=3, padding="same"),
            layers.Conv2D(6, kernel_size=3, padding="same"),
            layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
            layers.Conv2D(7, kernel_size=3, padding="same"),
            layers.Conv2D(8, kernel_size=3, padding="same"),
            layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
            layers.Conv2D(9, kernel_size=3, padding="same"),
            layers.Conv2D(10, kernel_size=3, padding="same"),
            layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
            layers.Conv2D(11, kernel_size=3, padding="same"),
            layers.Conv2D(12, kernel_size=3, padding="same"),
            layers.Flatten(),
            layers.Dense(num_classes, activation="softmax")
        ]
    )
    
    return model
def get_model2(input_shape, num_classes):
    model = keras.Sequential(
        [
                layers.Conv2D(12, kernel_size=3, padding="same", input_shape=input_shape),
                layers.Conv2D(11, kernel_size=3, padding="same"),
                layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
                layers.Conv2D(10, kernel_size=3, padding="same"),
                layers.Conv2D(9, kernel_size=3, padding="same"),
                layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
                layers.Conv2D(8, kernel_size=3, padding="same"),
                layers.Conv2D(7, kernel_size=3, padding="same"),
                layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
                layers.Conv2D(6, kernel_size=3, padding="same"),
                layers.Conv2D(5, kernel_size=3, padding="same"),
                layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
                layers.Conv2D(4, kernel_size=3, padding="same"),
                layers.Conv2D(3, kernel_size=3, padding="same"),
                layers.Flatten(),
                layers.Dense(num_classes, activation="softmax")
        ]

    )

    return model
def get_model3(input_shape, num_classes):
    model = keras.Sequential(
        [
                layers.Conv2D(3, kernel_size=3, padding="same", input_shape=input_shape),
                layers.Conv2D(4, kernel_size=3, padding="same"),
                layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
                layers.Conv2D(5, kernel_size=3, padding="same"),
                layers.Conv2D(6, kernel_size=3, padding="same"),
                layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
                layers.Conv2D(7, kernel_size=3, padding="same"),
                layers.Conv2D(8, kernel_size=3, padding="same"),
                layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
                layers.Conv2D(9, kernel_size=3, padding="same"),
                layers.Conv2D(10, kernel_size=3, padding="same"),
                layers.MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same"),
                layers.Conv2D(11, kernel_size=3, padding="same"),
                layers.Conv2D(12, kernel_size=3, padding="same"),
                layers.Flatten(),
                layers.Dense(num_classes, activation="softmax")
        ]
    )

    
    return model      

#Model 1 

In [19]:
rows, col = 28, 28 
input_shape = (1, rows, col)
epochs = 25
verbose = 0
optimizer = SGD()
model = get_model1(input_shape, num_classes)
model.summary()
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
model.fit(x_train, Y_train, batch_size=256, epochs=epochs, validation_split=0.2)
result = model.evaluate(x_test, Y_test, verbose=verbose)
print('Accuracy:', result[1])


Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_63 (Conv2D)          (None, 1, 28, 3)          759       
                                                                 
 conv2d_64 (Conv2D)          (None, 1, 28, 4)          112       
                                                                 
 max_pooling2d_26 (MaxPoolin  (None, 1, 14, 4)         0         
 g2D)                                                            
                                                                 
 conv2d_65 (Conv2D)          (None, 1, 14, 5)          185       
                                                                 
 conv2d_66 (Conv2D)          (None, 1, 14, 6)          276       
                                                                 
 max_pooling2d_27 (MaxPoolin  (None, 1, 7, 6)          0         
 g2D)                                                 

#Model 2 

In [20]:
rows, col = 28, 28
input_shape = (1, rows, col)
epochs = 25
verbose = 1
optimizer = Adam()
model = get_model2(input_shape, num_classes)
model.summary()
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
model.fit(x_train, Y_train, batch_size=64, epochs=epochs, validation_split=0.2)
result = model.evaluate(x_test, Y_test, verbose=verbose)
print('Accuracy:', result[1])

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_73 (Conv2D)          (None, 1, 28, 12)         3036      
                                                                 
 conv2d_74 (Conv2D)          (None, 1, 28, 11)         1199      
                                                                 
 max_pooling2d_30 (MaxPoolin  (None, 1, 14, 11)        0         
 g2D)                                                            
                                                                 
 conv2d_75 (Conv2D)          (None, 1, 14, 10)         1000      
                                                                 
 conv2d_76 (Conv2D)          (None, 1, 14, 9)          819       
                                                                 
 max_pooling2d_31 (MaxPoolin  (None, 1, 7, 9)          0         
 g2D)                                                 

#Model 3 

In [22]:
rows, col = 28, 28
input_shape = (1, rows, col)
epochs = 10
verbose = 1
optimizer = SGD()
model = get_model3(input_shape, num_classes)
model.summary()
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
model.fit(x_train, Y_train, batch_size=64, epochs=epochs, validation_split=0.2)
result = model.evaluate(x_test, Y_test, verbose=verbose)
print('Accuracy:', result[1])

Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_93 (Conv2D)          (None, 1, 28, 3)          759       
                                                                 
 conv2d_94 (Conv2D)          (None, 1, 28, 4)          112       
                                                                 
 max_pooling2d_38 (MaxPoolin  (None, 1, 14, 4)         0         
 g2D)                                                            
                                                                 
 conv2d_95 (Conv2D)          (None, 1, 14, 5)          185       
                                                                 
 conv2d_96 (Conv2D)          (None, 1, 14, 6)          276       
                                                                 
 max_pooling2d_39 (MaxPoolin  (None, 1, 7, 6)          0         
 g2D)                                                

#**Question 2**

In [23]:
from keras import backend as K
from keras.models import Sequential
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D
from keras.layers.core import Activation
from keras.layers.core import Flatten
from keras.layers.core import Dense
from keras.datasets import cifar10 
from keras.utils import np_utils
from tensorflow.keras.optimizers import SGD, RMSprop, Adam
import numpy as np

#the ConvNet
@staticmethod
def build(input_shape, classes):
    model = keras.Sequential(
        [
            layers.Conv2D(6, kernel_size=5, padding="same", input_shape=input_shape),
            layers.Activation("relu"),
            layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)),
            layers.Conv2D(16, kernel_size=5, padding="same"),
            layers.Activation("relu"),
            layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)),
            layers.Conv2D(120, kernel_size=5, padding="same"),
            layers.Activation("relu"),
            layers.Flatten(),
            layers.Dense(84),
            layers.Activation("relu"),
            layers.Dense(classes, activation="softmax")
        ]
    )
        
    return model

In [24]:
epochs = 25
batch_size = 128
verbose = 1
optimizer = Adam()
num_classes = 10
input_shape = (32,32,3)
(train_images, train_labels), (test_images, test_labels) = cifar10.load_data()

In [25]:
train_images = train_images.astype('float32')
test_images = test_images.astype('float32')
train_images = train_images / 255
test_images = test_images / 255 
num_classes = 10
train_labels = np_utils.to_categorical(train_labels, num_classes)
test_labels = np_utils.to_categorical(test_labels, num_classes)

#1.	What is the effect of learning rate on the training process? Which performed best?  

In [26]:
lr_test_result = []
lr_result = []
for lr in [0.1,0.01,0.25,0.5,0.001]:
		model = build(input_shape=input_shape, classes=num_classes)
		model.compile(loss="categorical_crossentropy", optimizer=Adam(learning_rate=lr), metrics=["accuracy"])
		history = model.fit(train_images, train_labels, batch_size=batch_size, epochs=epochs, verbose=verbose, validation_split=0.2)
		result = model.evaluate(test_images, test_labels, verbose=verbose)
		lr_test_result.append(result[1])
		lr_result.append(history.history['accuracy'][-1])
print(lr_result)
print(lr_test_result)

Epoch 1/25
313/313 [==============================] - 16s 50ms/step - loss: 227.0962 - accuracy: 0.0995 - val_loss: 2.3061 - val_accuracy: 0.1022
Epoch 2/25
313/313 [==============================] - 15s 49ms/step - loss: 2.3077 - accuracy: 0.0986 - val_loss: 2.3063 - val_accuracy: 0.0997
Epoch 3/25
313/313 [==============================] - 15s 49ms/step - loss: 2.3084 - accuracy: 0.1014 - val_loss: 2.3079 - val_accuracy: 0.0977
Epoch 4/25
313/313 [==============================] - 15s 48ms/step - loss: 2.3079 - accuracy: 0.1027 - val_loss: 2.3104 - val_accuracy: 0.1014
Epoch 5/25
313/313 [==============================] - 15s 48ms/step - loss: 2.3092 - accuracy: 0.0984 - val_loss: 2.3066 - val_accuracy: 0.1016
Epoch 6/25
313/313 [==============================] - 15s 48ms/step - loss: 2.3080 - accuracy: 0.1014 - val_loss: 2.3044 - val_accuracy: 0.1014
Epoch 7/25
313/313 [==============================] - 15s 48ms/step - loss: 2.3092 - accuracy: 0.1004 - val_loss: 2.3053 - val_accurac

#2.	What is the effect of batch_size size on the training process? Which performed best?  

In [28]:
bs_result = []
bs_test_result = []
for i in [8,16,32,64,128,256]:
    model = build(input_shape=input_shape, classes=num_classes)
    model.compile(loss="categorical_crossentropy", optimizer=Adam(learning_rate=0.1),
    metrics=["accuracy"])
    history = model.fit(train_images, train_labels,
    batch_size=i, epochs=epochs,
    verbose=verbose,  validation_split=0.2)
    result = model.evaluate(test_images, test_labels, verbose=verbose)
    bs_result.append(history.history['accuracy'][-1])
    bs_test_result.append(result[1])
print(bs_result)
print(bs_test_result)

Epoch 1/25
5000/5000 [==============================] - 41s 8ms/step - loss: 24.5857 - accuracy: 0.0995 - val_loss: 2.3257 - val_accuracy: 0.0952
Epoch 2/25
5000/5000 [==============================] - 41s 8ms/step - loss: 2.3280 - accuracy: 0.1010 - val_loss: 2.3278 - val_accuracy: 0.1003
Epoch 3/25
5000/5000 [==============================] - 41s 8ms/step - loss: 2.3289 - accuracy: 0.0981 - val_loss: 2.3303 - val_accuracy: 0.0980
Epoch 4/25
5000/5000 [==============================] - 41s 8ms/step - loss: 2.3284 - accuracy: 0.1026 - val_loss: 2.3356 - val_accuracy: 0.0980
Epoch 5/25
5000/5000 [==============================] - 41s 8ms/step - loss: 2.3291 - accuracy: 0.0979 - val_loss: 2.3366 - val_accuracy: 0.0977
Epoch 6/25
5000/5000 [==============================] - 41s 8ms/step - loss: 2.3288 - accuracy: 0.0998 - val_loss: 2.3595 - val_accuracy: 0.1022
Epoch 7/25
5000/5000 [==============================] - 41s 8ms/step - loss: 2.3292 - accuracy: 0.0997 - val_loss: 2.3375 - val_a

#3. Try different hyperparameters to obtain the best accuracy on the test set. What is your best performance and what were the hyperparameters?

In [36]:
epo_result = []
epo_test_result = []
for i in [5,10,15,20]:
    # initialize the optimizer and model
    model = build(input_shape=input_shape, classes=num_classes)
    model.compile(loss="categorical_crossentropy", optimizer=Adam(learning_rate=0.1),
    metrics=["accuracy"])
    history = model.fit(train_images, train_labels,
    batch_size=64, epochs=i,
    verbose=verbose,  validation_split=0.2)
    result = model.evaluate(test_images, test_labels, verbose=verbose)
    epo_result.append(history.history['accuracy'][-1])
    epo_test_result.append(result[1])
print(epo_result)
print(epo_test_result)

Epoch 1/5
625/625 [==============================] - 18s 28ms/step - loss: 25.2105 - accuracy: 0.0991 - val_loss: 2.3096 - val_accuracy: 0.1014
Epoch 2/5
625/625 [==============================] - 18s 28ms/step - loss: 2.3116 - accuracy: 0.1003 - val_loss: 2.3111 - val_accuracy: 0.0980
Epoch 3/5
625/625 [==============================] - 18s 29ms/step - loss: 2.3120 - accuracy: 0.0971 - val_loss: 2.3181 - val_accuracy: 0.1003
Epoch 4/5
625/625 [==============================] - 18s 29ms/step - loss: 2.3109 - accuracy: 0.1010 - val_loss: 2.3158 - val_accuracy: 0.1025
Epoch 5/5
313/313 [==============================] - 2s 6ms/step - loss: 2.3077 - accuracy: 0.1000
Epoch 1/10
625/625 [==============================] - 18s 29ms/step - loss: 146.9606 - accuracy: 0.0980 - val_loss: 2.3160 - val_accuracy: 0.1025
Epoch 2/10
625/625 [==============================] - 17s 28ms/step - loss: 2.3117 - accuracy: 0.1003 - val_loss: 2.3078 - val_accuracy: 0.1014
Epoch 3/10
625/625 [==================

#4. Implement an equivalent feed forward network for the same task with each hidden layer containing the same number of neurons as the number of parameters in each convolution layer. Use the ‘Adam’ optimizer to train your network on the CIFAR-10 dataset for a fixed set of 25 epochs. Compare its performance with your LeNet implementation based on the following questions:
#a. What is its performance?
#b. How many parameters are there in this network compared to the LeNet implementation? Are they worth it?

In [29]:
import keras
from keras import layers

@staticmethod
def model_4(input_shape, classes):
        model = keras.Sequential(
            [
                layers.Conv2D(64, kernel_size=5, padding="same",input_shape=input_shape, activation="relu"),
                layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)),
                layers.Conv2D(64, kernel_size=5, padding="same", activation="relu"),
                layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)),
                layers.Conv2D(64, kernel_size=5, padding="same", activation="relu"),
                layers.Flatten(),
                layers.Dense(64, activation="relu"),
                layers.Dense(classes, activation="softmax")
            ]
        )

        return model

In [30]:
model = model_4(input_shape=input_shape, classes=num_classes)
model.compile(loss="categorical_crossentropy", optimizer=Adam(learning_rate=0.1),
metrics=["accuracy"])
history = model.fit(train_images, train_labels,
batch_size=64, epochs=25,
verbose=verbose,  validation_split=0.2)
result = model.evaluate(test_images, test_labels, verbose=verbose)
print("Training accuracy", history.history['accuracy'][-1])
print("Accuracy", result[1])

Epoch 1/25
625/625 [==============================] - 74s 118ms/step - loss: 3013.7437 - accuracy: 0.1012 - val_loss: 2.3118 - val_accuracy: 0.0977
Epoch 2/25
625/625 [==============================] - 74s 119ms/step - loss: 2.3108 - accuracy: 0.0988 - val_loss: 2.3059 - val_accuracy: 0.0997
Epoch 3/25
625/625 [==============================] - 74s 119ms/step - loss: 2.3101 - accuracy: 0.1006 - val_loss: 2.3066 - val_accuracy: 0.0952
Epoch 4/25
625/625 [==============================] - 73s 117ms/step - loss: 2.3104 - accuracy: 0.1010 - val_loss: 2.3131 - val_accuracy: 0.1014
Epoch 5/25
625/625 [==============================] - 73s 116ms/step - loss: 2.3118 - accuracy: 0.1014 - val_loss: 2.3068 - val_accuracy: 0.0997
Epoch 6/25
625/625 [==============================] - 74s 118ms/step - loss: 2.3123 - accuracy: 0.0981 - val_loss: 2.3055 - val_accuracy: 0.1025
Epoch 7/25
625/625 [==============================] - 73s 117ms/step - loss: 2.3113 - accuracy: 0.1011 - val_loss: 2.3124 - val

#**Question 3**

Consider the below matrices as input (X) and convolutional kernel f. Consider that the depth of the input is 1.

In [31]:
import cv2
import numpy as np

x = np.array([[[7], [5], [0], [0], [3], [2]],
                   [[6], [4], [5], [1], [4], [8]],
                   [[9], [0], [2], [2], [5], [4]],
                   [[6], [3], [4], [7], [9], [8]],
                   [[5], [7], [5], [6], [9], [0]],
                   [[7], [9], [0], [8], [2], [3]]])
x = x.astype('float32') 

f = np.array([[1, 0, -1],
                   [2, 0, -2],
                   [1, 0, -1]])


#1.	What are the dimensions of the input and the kernel (or filter)? How many parameters are there in the kernel f? 

In [32]:
print("Dimension of input",x.shape[0])
print("Dimension of filter",f.shape[0])
print("Parameters in kernal", f.shape[0]*f.shape[1])

Dimension of input 6
Dimension of filter 3
Parameters in kernal 9


#2.	What is the output activation map when you apply the convolutional operation using the filter f on the input X without padding? 

In [33]:
import keras.backend as K
import numpy as np
from keras import Input, layers
from keras.models import Model

def my_filter(shape, dtype=None):

    f = np.array([
            [[[1]], [[0]], [[-1]]],
            [[[2]], [[0]], [[-2]]],
            [[[1]], [[0]], [[-1]]]
        ])
    assert f.shape == shape
    return K.variable(f, dtype='float32')
X = np.array([
                   [[7], [5], [0], [0], [3], [2]],
                   [[6], [4], [5], [1], [4], [8]],
                   [[9], [0], [2], [2], [5], [4]],
                   [[6], [3], [4], [7], [9], [8]],
                   [[5], [7], [5], [6], [9], [0]],
                   [[7], [9], [0], [8], [2], [3]]])

X = X.reshape((1, 6, 6, 1))
def build_model():
    input_tensor = Input(shape=(6,6,1))

    x = layers.Conv2D(filters=1, 
                      kernel_size = 3,
                      kernel_initializer=my_filter,
                      strides=1, 
                      padding='valid') (input_tensor)

    model = Model(inputs=input_tensor, outputs=x)
    return model
model = build_model()
out = model.predict(X)
print(out)

1/1 [==============================] - 0s 91ms/step
[[[[ 16.]
   [  9.]
   [ -4.]
   [-18.]]

  [[ 17.]
   [ -5.]
   [-10.]
   [-12.]]

  [[ 11.]
   [ -9.]
   [-17.]
   [  2.]]

  [[  9.]
   [ -1.]
   [-15.]
   [ 16.]]]]


#3.	What is the output when you apply a max-pooling operation on the output from the previous question? 

In [34]:
import numpy as np
from keras.models import Sequential 
from keras.layers import MaxPooling2D 

model = Sequential( 
    [MaxPooling2D(pool_size = 3, strides = 1)]) 
output = model.predict(out)   
output = np.squeeze(output) 
print(output) 

1/1 [==============================] - 0s 32ms/step
[[17.  9.]
 [17. 16.]]


In [35]:
model = build(input_shape=input_shape, classes=num_classes)
model.summary()

Model: "sequential_22"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_134 (Conv2D)         (None, 32, 32, 6)         456       
                                                                 
 activation_41 (Activation)  (None, 32, 32, 6)         0         
                                                                 
 max_pooling2d_63 (MaxPoolin  (None, 16, 16, 6)        0         
 g2D)                                                            
                                                                 
 conv2d_135 (Conv2D)         (None, 16, 16, 16)        2416      
                                                                 
 activation_42 (Activation)  (None, 16, 16, 16)        0         
                                                                 
 max_pooling2d_64 (MaxPoolin  (None, 8, 8, 16)         0         
 g2D)                                                

Question 1
1.) Learning Rate = 0.1, batch_size =256, epochs = 25, accuracy 0.113
2.) Learning Rate = 0.1, batch_size =64, epochs = 25, accuracy 0.113
3.) Learning Rate = 0.1, batch_size =64, epochs = 25, accuracy 0.186

Question 2
1.) Learning rate = [0.1,0.01,0.25,0.5,0.001], batch size = 128, epochs = 25, Accuracy decresed with increasing leraning rate, best performing = 0.01
2.) Batch Size = [8,16,32,64,128,256], Learning rate = 0.1, epochs = 25, Accuracy varied with batch size but not definitevly, best performing = 64,128
2.) epochs = [5,10,15,20], Learning rate = 0.1, batch size = 64, Accuracy varied with epochs but not definitevly, best performing = 10
4.)a.)Accuracy = 0.100
b.) There were more hyperparameters comparatively, no it took nore time to execute relatively.

References:
1.) https://keras.io/examples/vision/mnist_convnet/
2.) https://keras.io/api/datasets/mnist/#load_data-function
3.) https://keras.io/api/datasets/cifar10/
4.) https://github.com/sushantkumar-estech/LeNet-CNN-for-CIFAR-10-Classification-Dataset/blob/master/LeNet_Convolutional_Neural_Network_for_CIFAR_10_Classification_Dataset.ipynb